rmw_iceoryx_cpp/src/rmw_service_server_is_available.cpp

```diff
@@ -29,7 +32,29 @@ rmw_service_server_is_available(
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_ERROR);
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(is_available, RMW_RET_ERROR);
 
-  RMW_SET_ERROR_MSG("rmw_iceoryx_cpp does not support services.");
-  return RMW_RET_UNSUPPORTED;
+  rmw_ret_t ret = RMW_RET_ERROR;
+
+  auto iceoryx_client_abstraction = static_cast<IceoryxClient *>(client->data);
+  if (!iceoryx_client_abstraction) {
+    RMW_SET_ERROR_MSG("client data is null");
+    ret = RMW_RET_ERROR;
+    return ret;
+  }
+
+  auto iceoryx_client = iceoryx_client_abstraction->iceoryx_client_;
+  if (!iceoryx_client) {
+    RMW_SET_ERROR_MSG("iceoryx_client is null");
+    ret = RMW_RET_ERROR;
+    return ret;
+  }
+
+  std::this_thread::sleep_for(std::chrono::seconds(1));
```

Comment:
One second seemed like a good magic number to me. I can change this to 42 seconds, though :grin:

Reply:
Hahaha, in this case I vote for 73 ;)

Joking aside, why is the sleep necessary?

Reply:
For the discovery loop to have completed its run. As far as I know, this method is called for the creation of the ROS graph and I also remember there were some workarounds in ROS 2 needed for some buggy behaviour. As I'm a chicken I'll leave it as 1s for now.

See here for the explanation of the method: https://docs.ros.org/en/humble/p/rmw/generated/function_rmw_8h_1af0cec9687e5e4738aba7e061014eb3d5.html

Reply:
Oh, in this case you could evaluate whether the `iox::roudi::DISCOVERY_INTERVAL` from `iceoryx_posh_types.hpp` would work. In theory this should be the case ... maybe times two to satisfy the inner chicken :)

Oh, I just had an idea. In some examples we just defined a timeout and periodically checked the connection state until either the timeout expired or the expecdted state was reached. Similar to this code (may not compile and you might fix some `!` or `&&`)
```cpp
auto timeout = iox::deadline_timer(n * iox::roudi::DISCOVERY_INTERVAL);
*is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
while(!timeout.hasExpired() && !*is_available) {
    std::this_thread::sleep_for(iox::roudi::DISCOVERY_INTERVAL);
    *is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
}
```
This will massively speed up anything which uses this code, especially when the server is already available. Imagine someone creates multiple clients or maybe it is just a precondition in a loop before the client sends a request, then the unconditional sleep would make the call unbearable slow.

The documentation of `rmw_service_server_is_available` also does not suggest that something will artificially delay this call so it might be surprising for the user. The more I think about it, the more I would even fully remove the sleep since the call states that only the services discovered so far by the local node will be considered.

Reply:
Oh, it just occurred to me that the proposal with the while loop might not work in cases the user expects the service to not be available. Not sure in which situation this is used generally.

But I have another request. Independent of my proposal above, could you please just use
```cpp
*is_available = iceoryx_client->getConnectionState() == iox::ConnectionState::CONNECTED;
```
instead of the if-statement?
According to the documentation `is_available` is either `true` or `false` but currently it is only set to `true` in the if-statement. Now let's assume you have a user with this code, which is valid according the documentation
```cpp
boo is_available;
if(rmw_service_server_is_available(node_ptr, client_ptr, &is_available) == RMW_RET_OK && is_available) {
    // do something
}
```
Now some undefined behavior is introduced since `is_available` might not be set. If the compiler inlines `rmw_service_server_is_available` the whole if-statement could be removed and `is_available` be set unconditionally to `true` since not entering the if statement would cause undefined behaviour and since the compiler can assume the code does not execute undefined behaviour, the if-statement must always be `true` ... I guess there is a reason I don't like out-parameter :D